Jetsnack/app/src/main/java/com/example/jetsnack/ui/JetsnackApp.kt

```diff
@@ -17,30 +17,155 @@
-            val tabs = remember { HomeSections.values() }
-            val navController = rememberNavController()
+            val scaffoldState = rememberScaffoldState()
+            val scaffoldStateHolder = rememberScaffoldStateHolder(scaffoldState)
+
             JetsnackScaffold(
-                bottomBar = { JetsnackBottomBar(navController = navController, tabs = tabs) }
+                scaffoldState = scaffoldState,
+                bottomBar = {
+                    JetsnackBottomBar(scaffoldStateHolder.navController, scaffoldStateHolder.tabs)
+                }
             ) { innerPaddingModifier ->
                 JetsnackNavGraph(
-                    navController = navController,
+                    scaffoldStateHolder = scaffoldStateHolder,
                     modifier = Modifier.padding(innerPaddingModifier)
                 )
             }
         }
     }
 }
+
+/**
+ * Creates a [ScaffoldStateHolder] and memoizes it.
+ *
+ * Pending snackbar messages to show on the screen are remembered across
+ * activity and process recreation.
+ */
+@Composable
+private fun rememberScaffoldStateHolder(
+    scaffoldState: ScaffoldState = rememberScaffoldState()
+): ScaffoldStateHolder {
+    val navController = rememberNavController()
+    val coroutineScope = rememberCoroutineScope()
+    val lifecycle = LocalLifecycleOwner.current.lifecycle
+    val snackbarHostState = scaffoldState.snackbarHostState
+
+    return rememberSaveable(
+        inputs = arrayOf(navController, snackbarHostState, coroutineScope, lifecycle),
+        saver = ScaffoldStateHolderSaver(
+            navController, snackbarHostState, coroutineScope, lifecycle
+        )
+    ) {
+        ScaffoldStateHolder(
+            navController, scaffoldState.snackbarHostState, coroutineScope, lifecycle, listOf()
+        )
+    }
+}
+
+/**
+ * State holder for Jetsnack's app scaffold that contains state related to [JetsnackScaffold], and
+ * handles Navigation and Snackbar events.
+ */
+class ScaffoldStateHolder(
+    val navController: NavHostController,
+    private val snackbarHostState: SnackbarHostState,
+    coroutineScope: CoroutineScope,
+    lifecycle: Lifecycle,
+    initialSnackbarMessages: List<String>
+) {
+    // Tabs for JetsnackBottomBar
+    val tabs = HomeSections.values()
+
+    // Queue a maximum of 3 snackbar messages
+    private val snackbarMessages = Channel<String>(3, BufferOverflow.DROP_OLDEST)
+    // Cache of the Channel buffer used to restore state
+    private val pendingSnackbarMessages = mutableListOf<String>()
+
+    init {
+        // Restore snackbar messages state
+        initialSnackbarMessages.forEach { showSnackbar(it) }
+
+        // Process snackbar events only when the lifecycle is at least STARTED
+        coroutineScope.launch {
+            snackbarMessages
+                .receiveAsFlow()
+                .onEach { pendingSnackbarMessages.remove(it) }
+                .flowWithLifecycle(lifecycle)
```

Comment:
I don't think this behavior is quite what is desired.

`flowWithLifecycle` uses `callbackFlow` under the hood, which by default has a buffer size of `Channel.BUFFERED` (which is 64 by default).

So even though the below `collect` is slow (since `snackbarHostState.showSnackbar` suspends until the message is shown and then hidden), up to 64 messages will be buffered and effectively enqueued to be displayed.

Adding a `.buffer(Channel.RENDEZVOUS)` after `flowWithLifecycle` will remove that 64 element buffer, so the desired 3 element buffer of the `snackbarMessages` `Channel` will be used instead.

Reply:
Hmm, upon further investigation, even `.buffer(Channel.RENDEZVOUS)` will allow an unwanted element to get through with `flowWithLifecycle`:

Suppose `showSnackbar` is called with successive numbers `1`, `2`, `3`, `4`, `5`, ..., `9`, `10` extremely quickly (while the snackbar for `1` is showing)

`2` will be displayed next, followed by `8`, `9`, and `10`, when we really just want `8`, `9` and `10`. I think that is because `2` is internally received by `flowWithLifecycle` even though `1` is still being displayed.

Notably, using `repeatOnLifecycle` directly can avoid that, resulting in the desired `1`, `8`, `9`, `10`:

```kotlin
        coroutineScope.launch {
            lifecycle.repeatOnLifecycle(Lifecycle.State.STARTED) {
                snackbarMessages
                    .receiveAsFlow()
                    .collect { message ->
                        snackbarHostState.showSnackbar(message)
                    }
            }
        }
```

The difference is subtle, but it ensures that a message is received from the channel if and only the `snackbarHostState.showSnackbar` is called with it.

Reply:
Good catch! The underlying implementation of `callbackFlow` bits again :) thank you!